application/src/test/java/org/thingsboard/server/transport/coap/rpc/AbstractCoapServerSideRpcIntegrationTest.java

```diff
@@ -94,23 +104,38 @@ protected void processTwoWayRpcTest(String expectedResponseResult, boolean proto
         CoapTestCallback callbackCoap = new TestCoapCallbackForRPC(client, 1, false, protobuf);
 
         CoapObserveRelation observeRelation = client.getObserveRelation(callbackCoap);
-        callbackCoap.getLatch().await(3, TimeUnit.SECONDS);
-
+        String awaitAlias = "await Two Way Rpc (client.getObserveRelation)";
+        await(awaitAlias)
+                .atMost(10, TimeUnit.SECONDS)
+                .until(() -> CoAP.ResponseCode.VALID.equals(callbackCoap.getResponseCode()) &&
+                        callbackCoap.getObserve() != null &&
+                        0 == callbackCoap.getObserve().intValue());
         validateCurrentStateNotification(callbackCoap);
 
         String setGpioRequest = "{\"method\":\"setGpio\",\"params\":{\"pin\": \"26\",\"value\": 1}}";
         String deviceId = savedDevice.getId().getId().toString();
-
+        int expectedObserveBeforeGpioRequestAddCnt1 = callbackCoap.getObserve().intValue() + 1;
         String actualResult = doPostAsync("/api/rpc/twoway/" + deviceId, setGpioRequest, String.class, status().isOk());
-        callbackCoap.getLatch().await(3, TimeUnit.SECONDS);
+        awaitAlias = "await Two Way Rpc (setGpio(method, params, value) first";
+        await(awaitAlias)
+                .atMost(10, TimeUnit.SECONDS)
+                .until(() -> CoAP.ResponseCode.CONTENT.equals(callbackCoap.getResponseCode()) &&
+                        callbackCoap.getObserve() != null &&
+                        expectedObserveBeforeGpioRequestAddCnt1 == callbackCoap.getObserve().intValue());
+        validateTwoWayStateChangedNotification(callbackCoap, expectedResponseResult, actualResult);
 
-        validateTwoWayStateChangedNotification(callbackCoap, 1, expectedResponseResult, actualResult);
+        validateTwoWayStateChangedNotification(callbackCoap, expectedResponseResult, actualResult);
```

Comment:
Hi @nickAS21, why do you do the same validation twice?

Reply:
OK